### ISSUE TITLE
Fail to parse StoreKitTest receipt

### ISSUE THREAD

**Issue by motocodeltd** [tikhop/TPInAppReceipt#58]
### Platform

- [X ] iOS
- [ ] macOS
- [ ] tvOS

### Environment

- [ X] Sandbox
- [ ] Production

### Version
ℹ Please replace this with the version of TPInAppReceipt you're using.
16ac80c7f2cf0810dbe1de0c5ac48ac67b031d0b (latest master at time of writing)

## Report
Setting up a test using 
https://developer.apple.com/documentation/xcode/setting_up_storekit_testing_in_xcode - this is a manual test, not using StoreKitTest

InAppReceipt.local() fails because pkcs7.extractInAppPayload()! returns Nil (InAppReceipt:67).

I'm aware of the different cert, but I don't even get to the point of loading the cert

### Issue summary
Set up your StoreKitTest file. Upgrade. Validate the receipt.

### What did you expect to happen
receipt should be parsed.

### What happened instead
crash due to force unwrap of nil

**Comment by motocodeltd** [comment]
MIAGCSqGSIb3DQEHAqCAMIACAQExDzANBglghkgBZQMEAgEFADCABgkqhkiG9w0BBwGggCSABIIBfzGCAXswDwIBAAIBAQQHDAVYY29kZTALAgEBAgEBBAMCAQAwIwIBAgIBAQQbDBluZXQuemFjaGFyaWFkaXMuY3ljbGVtYXBzMBECAQMCAQEECQwHMzEuMTAuMDAQAgEEAgEBBAi9dvT/BAAAADAcAgEFAgEBBBQNsT28UFTnpyaDfWk3iEEyTGZufTAKAgEIAgEBBAIWADAiAgEMAgEBBBoWGDIwMjAtMDctMjJUMTg6MzM6MTUrMDEwMDCBngIBEQIBAQSBlTGBkjAMAgIGpQIBAQQDAgEBMBwCAgamAgEBBBMMEUNZQ0xFTUFQU19QUkVNSVVNMAwCAganAgEBBAMMATAwIwICBqgCAQEEGhYYMjAyMC0wNy0yMlQxODozMzoxNCswMTAwMCMCAgasAgEBBBoWGDIwMjEtMDctMjJUMTg6MzM6MTQrMDEwMDAMAgIGtwIBAQQDAgEAMCICARUCAQEEGhYYNDAwMS0wMS0wMVQwMDowMDowMCswMDAwAAAAAAAAoIIDeDCCA3QwggJcoAMCAQICAQEwDQYJKoZIhvcNAQELBQAwXzERMA8GA1UEAwwIU3RvcmVLaXQxETAPBgNVBAoMCFN0b3JlS2l0MREwDwYDVQQLDAhTdG9yZUtpdDELMAkGA1UEBhMCVVMxFzAVBgkqhkiG9w0BCQEWCFN0b3JlS2l0MB4XDTIwMDQwMTE3NTIzNVoXDTQwMDMyNzE3NTIzNVowXzERMA8GA1UEAwwIU3RvcmVLaXQxETAPBgNVBAoMCFN0b3JlS2l0MREwDwYDVQQLDAhTdG9yZUtpdDELMAkGA1UEBhMCVVMxFzAVBgkqhkiG9w0BCQEWCFN0b3JlS2l0MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA23+QPCxzD9uXJkuTuwr4oSE+yGHZJMheH3U+2pPbMRqRgLm/5QzLPLsORGIm+gQptknnb+Ab5g1ozSVuw3YI9UoLrnp0PMSpC7PPYg/7tLz324ReKOtHDfHti6z1n7AJOKNue8smUAoa4YnRcnYLOUzLT27As1+3lbq5qF1KdKvvb0GlfgmNuj09zXBX2O3v1dp3yJMEHO8JiHhlzoHyjXLnBxpuJhL3MrENuziQawbE/A3llVDNkci6JfRYyYzhcdtKRfMtGZYDVoGmRO51d1tTz3isXbo+X1ArXCmM3cLXKhffIrTX5Hior6htp8HaaC1mzM8pC1As48L75l8SwQIDAQABozswOTAPBgNVHRMBAf8EBTADAQH/MA4GA1UdDwEB/wQEAwIChDAWBgNVHSUBAf8EDDAKBggrBgEFBQcDAzANBgkqhkiG9w0BAQsFAAOCAQEAsgDgPPHo6WK9wNYdQJ5XuTiQd3ZS0qhLcG64Z5n7s4pVn+8dKLhfKtFznzVHN7tG03YQ8vBp7M1imXH5YIqESDjEvYtnJbmrbDNlrdjCmnhID+nMwScNxs9kPG2AWTOMyjYGKhEbjUnOCP9mwEcoS+tawSsJViylqgkDezIx3OiFeEjOwMUSEWoPDK4vBcpvemR/ICx15kyxEtP94x9eDX24WNegfOR/Y6uXmivDKtjQsuHVWg05G29nKKkSg9aHeG2ZvV6zCuCYzvbqw45taeu3QIE9hz1wUdHEXY2l3H9qWBreYHY3Uuz/rBldDBUvig/1icjXKx0e7CuRBac9TzGCAY8wggGLAgEBMGQwXzERMA8GA1UEAwwIU3RvcmVLaXQxETAPBgNVBAoMCFN0b3JlS2l0MREwDwYDVQQLDAhTdG9yZUtpdDELMAkGA1UEBhMCVVMxFzAVBgkqhkiG9w0BCQEWCFN0b3JlS2l0AgEBMA0GCWCGSAFlAwQCAQUAMA0GCSqGSIb3DQEBAQUABIIBADCN/ADgSni8q/ggISEZPfImgrYLUNvfp/uaV2nbTQl9BpJUzWoppyTi/NEndik15IybDhUXbxe8S2waAFfVK7VG3kT/GwGFwxxSiZfXlcs+41QkebkooufYnzw1JUmTOhrr7hvzGuOZ+RIX6HoIE/SjPgTFOJeCTxCYsr8pHaD78zM2Y6TSwtsV2xznO4CNoqS8KPOycK6Br0DC5g+dushclAOA4CLVAQpiT/qFQdFZwyg5svvLtIGci8SLRdzcgfK8Uu4GNOTygyZgGtTYrwhOkH6PJeRnnYhte+ocguyopHqpT7gtsHIeL/cFOSQXMy7ycT1h/gPrhWe+hDoteJsAAAAAAAA=

**Comment by motocodeltd** [comment]
this is the base64encoded receipt

**Comment by motocodeltd** [comment]
Thank you!

**Comment by tikhop** [comment]
Hi @motocodeltd,

I think this issue is related to https://github.com/tikhop/TPInAppReceipt/issues/55. 

I'm planing to work on it during the weekend.

Thanks,
P

**Comment by motocodeltd** [comment]
Thanks Pavel - will be interested in what you find out.

https://developer.apple.com/forums/thread/655313 <---- this is mirroring my experience too.

**Comment by tikhop** [comment]
Hi @motocodeltd,

First, I was wrong that it's related to #55. 
Second, I've tried to reproduce the issue using your receipt and found a problem with asn1 decoder. The structure of your receipt (I assume you got it by using new xcode and iOS 14?!) is quite different compare to the "Original" format.

I'm working on it now. 

Thanks,
P

**Comment by motocodeltd** [comment]
Thanks for the update - yes this is a receipt via xcode 12 and ios14.

**Comment by holtwick** [comment]
Thanks @tikhop for looking into it! I experienced the same issue and can help testing. I use Xcode 12b3 on macOS Big Sur (Intel) for a macOS only app. The parsing indeed fails. You might also need the certificate as well. In case I attach it to this message (hoping it is the same for everyone and not generated locally)

<img width="1247" alt="Screenshot 2020-07-31 at 13 42 52" src="https://user-images.githubusercontent.com/49557/89031929-c7f90500-d333-11ea-9d63-b3dc2b29eb5a.png">

[StoreKitTestCertificate.cer.zip](https://github.com/tikhop/TPInAppReceipt/files/5006822/StoreKitTestCertificate.cer.zip)

**Comment by tikhop** [comment]
Hi @holtwick,

Thanks for testing. Currently, I'm writing new asn1 decoder in order to parse any ASN1 notations since in ios 14 we have a slightly different format of the receipt. It's still pkcs7, but with a different encoding technique. 

Another issue you are talking about is related to certificate chain and I'm aware about it as well. It shouldn't be hard to implement when I'm done with ASN1 decoder. 

I'm planing to get it done by the end of the week. 

Thanks,
P

**Comment by tikhop** [comment]
Alright, I'm done with ASN.1 Decoder (still have some minor issues, but in general it works well and parses both, legacy and new receipt formats). Here is the link, if someone wants to check it out. [ASN1Swift](https://github.com/tikhop/ASN1Swift). It works similar to `JSONDecoder` and implements the same protocol `Decoder`. Basically, you just provide an object with `CodingKeys` and `ASN1Decoder` deserialize `Data` into provided `Type`.

I've started migration from old decoder to the new one. It's on `develop` branch. I think I need a couple more days...

Best,
P

**Comment by tikhop** [comment]
@motocodeltd, @holtwick 

Please, for now, use the code from `develop` branch. Everything should work now. 

Thanks again for bringing this up, sharing a receipt and a certificate.

Best,
P

**Comment by holtwick** [comment]
Awesome, thanks a lot! I can confirm, that I do not get parsing errors any more. Sadly it is still not yet working as expected, but this might be due to some local configuration issues. I am under the impression it uses a wrong local receipt somehow, since the purchases I do in the StoreKit test environment are not reflected when I list the purchases found by TPInAppReceipt. I'll try to investigate further.

**Comment by holtwick** [comment]
>  I am under the impression it uses a wrong local receipt somehow

Indeed, if I turn StoreKit testing on/off it always uses the Receipt from the sandbox account purchases. Looks like I need to watch the WWDC videos again ;)

**Comment by tikhop** [comment]
Hey @holtwick,

I'm glad it parses the receipt at least. Unfortunately, I haven't got a chance to try it out neither with a real purchases nor with StoreKitTest. I just did it the way it says in the WWDC video. I will do more testing next week. Let me know if you have any other issues. 

Thanks

**Comment by motocodeltd** [comment]
Hi Pavel - this is working for me, thank you, after a couple of patches which I've submitted as PRs (#59 and #60 )

**Comment by tikhop** [comment]
Thanks for the patches.

**Comment by motocodeltd** [comment]
Hi Pavel - one more thing - are you expecting hash verification to work? signature, bundle identifier and version are working, but not hash.

Thank you

**Comment by tikhop** [comment]
Hi Stefanos, 

Actually it works for me and if you are testing on real device it suppose to work either. If you want to test in simulator instead, you have to change hardcoded guid for this purpose. It should match device's guid.

You can insert yours [here](https://github.com/tikhop/TPInAppReceipt/blob/8a93e65ba86678deb5fe9b0a87e8c28880724d6d/Sources/Validation.swift#L256)

**Comment by motocodeltd** [comment]
Hi Pavel, that's great, thank you. One more thing though, why is there a hardcoded uuid for the simulator? It works just fine if you remove the first part of the #if block and you use the uuid provided by the OS running on the simulator itself (at least, it works for me (tm) for watchos and ios). Not sure if I understand your reason for the hardcoded uuid, but I'll provide a PR that removes it anyway.

**Comment by tikhop** [comment]
I agree, it shouldn't be there. The only reason I put it there is in order to debug a receipt from a real device.

**Comment by holtwick** [comment]
Thanks @motocodeltd and @motocodeltd and @tikhop The fixes seem to have worked. I can now successfuly test on macOS Big Sur with Xcode beta! 🎉

**Comment by motocodeltd** [comment]
Hi Pavel - sorry to be a downer, but recent changes to ASN1 (post eb862bb7cd3a1a4e93d1914185d1053cfee66c5d) break receipt validation again. I got these transitively and randomly, as TPInAppReceipt depend on ASN1Swift master which is a moving target. I get assertionFailure("Something wrong here") InAppReceiptPayload. Sorry I can't be more helpful but I'm on holiday (!) and have little time to investigate further. Rolling back ASN1  to eb862bb7cd3a1a4e93d1914185d1053cfee66c5d fixes the issue. I think you can replicate with the receipt above.

Thanks again

**Comment by tikhop** [comment]
Got it. I will fix it in a bit.

Thanks for letting me know.

**Comment by tikhop** [comment]
Hey @motocodeltd, 

I've pushed some changes. Works well with my receipt and yours.

@holtwick would you mind to try the last changes and see whether it works well for you either. Thanks in advance.

**Comment by holtwick** [comment]
I have to admit I use the sources directly i.e. I copy the .swift files from both repos, modify some imports and then just use it. As of the latest from `develop` and ASN1 `master` it seems to work fine with my test case.

**Comment by Jerland2** [comment]
having this issue as well, is there a status on when this will be merged into master?

**Comment by tikhop** [comment]
Hi @Jerland2,

I've been thinking to do that when Apple release xcode 12 GM, but I think I will do it over the weekend. 

Thanks,
P

**Comment by chrisvanbuskirk** [comment]
The GM is out.

**Comment by tikhop** [comment]
@chrisvanbuskirk yep, merging in a couple hours...

**Comment by tikhop** [comment]
Just merged `develop` into `master` and released new cocoapods version. 

I'm closing this issue. 

Shout out to @motocodeltd, @holtwick, @Jerland2 and @chrisvanbuskirk for your help.

**Comment by chrisvanbuskirk** [comment]
@tikhop A little issue here with SPM

<img width="875" alt="Screen Shot 2020-09-16 at 10 03 27 AM" src="https://user-images.githubusercontent.com/3638007/93363167-1ab74f00-f804-11ea-8c23-b27fb6155334.png">

**Comment by tikhop** [comment]
@chrisvanbuskirk Fixed!
